File: dota-zone/dotazone/src/main/java/br/com/dotazone/model/service/ItemAsync.java
package br.com.dotazone.model.service;

import android.app.ProgressDialog;
import android.content.res.AssetManager;
import android.os.AsyncTask;
import android.support.v4.app.Fragment;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import br.com.dotazone.DotaZoneBrain;
import br.com.dotazone.R;
import br.com.dotazone.model.entity.Item;
import br.com.dotazone.model.entity.Item.ItemElementy;
import br.com.dotazone.model.entity.ItemAtrrib;
import br.com.dotazone.model.entity.ItemAtrrib.ItemAttribElementy;
import br.com.dotazone.model.util.UrlUtils;
import br.com.dotazone.view.activity.BaseActivity;
import br.com.dotazone.view.fragment.DialogError;
import br.com.dotazone.view.fragment.DialogError.TypeError;
import br.com.dotazone.view.fragment.ItemFragment;

public class ItemAsync extends AsyncTask<Void, Void, String> {

    private ProgressDialog mProgressDialog;
    private Fragment mFragment;
    private RequestREST mRequest;
    private AdapterAction mAdapterAction;
    private List<ItemAtrrib> mItemAttribs;
    private List<Item> mItems = new ArrayList<Item>();

    public ItemAsync(Fragment fragment) {

        this.mFragment = fragment;
        this.mRequest = new RequestREST();
        this.mProgressDialog = new ProgressDialog(fragment.getActivity());
        this.mAdapterAction = (ItemFragment) fragment;
    }

    @Override
    protected String doInBackground(Void... params) {
        String result = null;
        String resultAttrib = null;

        try {

            AssetManager assetManager = mFragment.getActivity().getAssets();
            InputStream inputStreamItemAttrib = assetManager.open("attribs_item.txt");
            resultAttrib = UrlUtils.convertStreamToString(inputStreamItemAttrib);
            result = mRequest.getHttpRequest(UrlUtils.getUrlItem(mFragment.getActivity()));
            if (DotaZoneBrain.items.isEmpty()) {
                createItemAttrib(resultAttrib);
                createItem(result);
            }

        } catch (Exception e) {

            Log.e(DotaZoneBrain.TAG, "Erro ao tentar consultar o arquivo de Item.--------");
            DialogError fragmentError = DialogError.newFragmentDialog(mFragment.getActivity().getResources()
                    .getString(R.string.error_item_connection), (BaseActivity) mFragment.getActivity(), TypeError.ONE_OPTIONS);
            // DotaZoneBrain.mItems = new ArrayList<Item>();
            fragmentError.show(mFragment.getActivity().getSupportFragmentManager(), "Item");
        }

        return result;
    }

    private void createItemAttrib(String result) {

        try {

            JSONObject jsonResponse = new JSONObject(result);
            JSONObject mainNode = (JSONObject) jsonResponse.get("itemdata");
            Iterator<?> keys = mainNode.keys();
            mItemAttribs = new ArrayList<ItemAtrrib>();
            while (keys.hasNext()) {
                String key = (String) keys.next();
                if (mainNode.get(key) instanceof JSONObject) {

                    ItemAtrrib itemAttrib = new ItemAtrrib();
                    JSONObject jsonChild = mainNode.getJSONObject(key);
                    try {
                        itemAttrib.setId(key);
                        itemAttrib.setDamage(jsonChild.getInt(ItemAttribElementy.DAMAGE.toString().toLowerCase()));
                        itemAttrib.setAgility(jsonChild.getInt(ItemAttribElementy.AGILITY.toString().toLowerCase()));
                        itemAttrib.setStrength(jsonChild.getInt(ItemAttribElementy.STRENGTH.toString().toLowerCase()));
                        itemAttrib.setInteligence(jsonChild.getInt(ItemAttribElementy.INTELIGENCE.toString().toLowerCase()));
                        itemAttrib.setArmor(jsonChild.getInt(ItemAttribElementy.ARMOR.toString().toLowerCase()));
                        itemAttrib.setMs(jsonChild.getString(ItemAttribElementy.MS.toString().toLowerCase()));

                        mItemAttribs.add(itemAttrib);
                    } catch (JSONException e) {

                        Log.e(DotaZoneBrain.TAG, "No item [" + itemAttrib.getId() + "  -]" + e);

                    }
                }
            }

        } catch (JSONException e) {
            DialogError fragmentError = DialogError.newFragmentDialog(mFragment.getActivity().getResources().getString(R.string.error_json_item),
                    (BaseActivity) mFragment.getActivity(), TypeError.ONE_OPTIONS);
            // DotaZoneBrain.mItems = new ArrayList<Item>();
            fragmentError.show(mFragment.getActivity().getSupportFragmentManager(), "Item");
        }

    }

    public void createItem(String result) {

        try {

            JSONObject jsonResponse = new JSONObject(result);
            JSONObject mainNode = (JSONObject) jsonResponse.get("itemdata");
            Iterator<?> keys = mainNode.keys();

            while (keys.hasNext()) {
                String key = (String) keys.next();
                if (mainNode.get(key) instanceof JSONObject) {

                    Item item = new Item();
                    JSONObject jsonChild = mainNode.getJSONObject(key);
                    JSONArray componentsList = null;
                    try {

                        if (jsonChild.getString(ItemElementy.DNAME.toString().toLowerCase()).equals("Dagon")) {
                            System.out.println();
                        }

                        item.setId(jsonChild.getInt(ItemElementy.ID.toString().toLowerCase()));
                        item.setDescription(jsonChild.getString(ItemElementy.DESC.toString().toLowerCase()));
                        item.setMc(jsonChild.getString(ItemElementy.MC.toString().toLowerCase()));
                        item.setCreated(jsonChild.getBoolean(ItemElementy.CREATED.toString().toLowerCase()));
                        item.setImageName(jsonChild.getString(ItemElementy.IMG.toString().toLowerCase()));
                        item.setQual(jsonChild.getString(ItemElementy.QUAL.toString().toLowerCase()));
                        item.setName(jsonChild.getString(ItemElementy.DNAME.toString().toLowerCase()));
                        item.setLore(jsonChild.getString(ItemElementy.LORE.toString().toLowerCase()));
                        item.setCreated(jsonChild.getBoolean(ItemElementy.CREATED.toString().toLowerCase()));
                        item.setCost(jsonChild.getInt(ItemElementy.COST.toString().toLowerCase()));
                        item.setNotes(jsonChild.getString(ItemElementy.NOTES.toString().toLowerCase()));
                        item.setAttribute(jsonChild.getString(ItemElementy.ATTRIB.toString().toLowerCase()));
                        item.setCloudown(jsonChild.getString(ItemElementy.CD.toString().toLowerCase()));

                        try {
                            componentsList = jsonChild.getJSONArray(ItemElementy.COMPONENTS.toString().toLowerCase());
                        } catch (Exception e) {

                            // Trata os itens que não tem um componente.
                            Log.i(DotaZoneBrain.TAG, "no improvement items[" + item.getName() + "  -]");
                        }

                        if (componentsList != null) {
                            for (int i = 0; i < componentsList.length(); i++) {
                                item.getComponents().add(componentsList.getString(i));
                                Log.i(DotaZoneBrain.TAG, "Improvement items[" + item.getName() + "  -]");

                            }
                        }

                        identifyItemCategoryAndAdd(item);

                    } catch (JSONException e) {

                        Log.e(DotaZoneBrain.TAG, "No item [" + item.getName() + "  -]" + e);

                    }

                }

            }

        } catch (JSONException e) {
            DialogError fragmentError = DialogError.newFragmentDialog(mFragment.getActivity().getResources().getString(R.string.error_json_item),
                    (BaseActivity) mFragment.getActivity(), TypeError.ONE_OPTIONS);
            // DotaZoneBrain.mItems = new ArrayList<Item>();
            fragmentError.show(mFragment.getActivity().getSupportFragmentManager(), "Item");
        }

    }

    private void identifyItemCategoryAndAdd(Item item) {

        int pos = item.getImageName().lastIndexOf('.');
        int id = mFragment.getActivity().getResources()
                .getIdentifier(item.getImageName().substring(0, pos), "drawable", mFragment.getActivity().getPackageName());

        for (ItemAtrrib itemAttrib : mItemAttribs) {
            if (item.getImageName().equals((itemAttrib.getId() + ".png"))) {

                item.setmItemAttrib(itemAttrib);
                Log.i(DotaZoneBrain.TAG, "attribute item add-" + itemAttrib.getId());
                break;
            }
        }

        // Verifica se existem a imagem cadastrada no json.
        if (id > 0) {

            DotaZoneBrain.items.add(item);
            //mItems.add(item);
            Log.i(DotaZoneBrain.TAG, "adding Item [" + item.getImageName() + "]");

        }

    }

    @Override
    protected void onPostExecute(String result) {
        super.onPostExecute(result);

        mProgressDialog.dismiss();
        mAdapterAction.initList();
    }

    @Override
    protected void onPreExecute() {

        mProgressDialog = new ProgressDialog(mFragment.getActivity());
        mProgressDialog.setMessage("Carregando...");
        mProgressDialog.show();
        mProgressDialog.setCancelable(false);
    }
}

File: app/src/main/java/com/br/dotazone/model/service/ItemAsync.kt
package com.br.dotazone.model.service

import android.app.ProgressDialog
import android.os.AsyncTask
import android.util.Log
import androidx.fragment.app.Fragment
import com.br.dotazone.DotaZoneBrain.TAG
import com.br.dotazone.DotaZoneBrain.items
import com.br.dotazone.R
import com.br.dotazone.domain.heroes.prov.Item
import com.br.dotazone.domain.heroes.prov.Item.ItemElementy
import com.br.dotazone.domain.heroes.prov.ItemAtrrib
import com.br.dotazone.domain.heroes.prov.ItemAtrrib.ItemAttribElementy
import com.br.dotazone.model.util.UrlUtils.Companion.convertStreamToString
import com.br.dotazone.model.util.UrlUtils.Companion.getUrlItem
import com.br.dotazone.view.activity.BaseActivity
import com.br.dotazone.view.fragment.DialogError.Companion.newFragmentDialog
import com.br.dotazone.view.fragment.DialogError.TypeError
import com.br.dotazone.view.fragment.ItemFragment
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.*


class ItemAsync(private val mFragment: Fragment) : AsyncTask<Void?, Void?, String?>() {
	private var mProgressDialog: ProgressDialog
	private val mRequest: RequestREST
	private val mAdapterAction: AdapterAction
	private var mItemAttribs: MutableList<ItemAtrrib> = mutableListOf()
	private val mItems: List<Item> = ArrayList()

	override fun doInBackground(vararg p0: Void?): String? {
		var result: String? = null
		var resultAttrib: String? = null
		try {
			val assetManager = mFragment.activity!!.assets
			val inputStreamItemAttrib = assetManager.open("attribs_item.txt")
			resultAttrib = convertStreamToString(inputStreamItemAttrib)
			result = mRequest.getHttpRequest(getUrlItem(mFragment.activity!!))
			if (items.isEmpty()) {
				createItemAttrib(resultAttrib)
				createItem(result)
			}
		} catch (e: Exception) {
			Log.e(TAG, "Erro ao tentar consultar o arquivo de Item.--------")
			val fragmentError = newFragmentDialog(mFragment.activity!!.resources
					.getString(R.string.error_item_connection), (mFragment.activity as BaseActivity?)!!, TypeError.ONE_OPTIONS)
			// DotaZoneBrain.mItems = new ArrayList<Item>();
			fragmentError.show(mFragment.activity!!.supportFragmentManager, "Item")
		}
		return result
	}

	private fun createItemAttrib(result: String) {
		try {
			val jsonResponse = JSONObject(result)
			val mainNode = jsonResponse["itemdata"] as JSONObject
			val keys: Iterator<*> = mainNode.keys()
			mItemAttribs = ArrayList()
			while (keys.hasNext()) {
				val key = keys.next() as String
				if (mainNode[key] is JSONObject) {
					val itemAttrib = ItemAtrrib()
					val jsonChild = mainNode.getJSONObject(key)
					try {
						itemAttrib.id = key
						itemAttrib.damage = jsonChild.getInt(ItemAttribElementy.DAMAGE.toString().toLowerCase())
						itemAttrib.agility = jsonChild.getInt(ItemAttribElementy.AGILITY.toString().toLowerCase())
						itemAttrib.strength = jsonChild.getInt(ItemAttribElementy.STRENGTH.toString().toLowerCase())
						itemAttrib.inteligence = jsonChild.getInt(ItemAttribElementy.INTELIGENCE.toString().toLowerCase())
						itemAttrib.armor = jsonChild.getInt(ItemAttribElementy.ARMOR.toString().toLowerCase())
						itemAttrib.ms = jsonChild.getString(ItemAttribElementy.MS.toString().toLowerCase())
						mItemAttribs.add(itemAttrib)
					} catch (e: JSONException) {
						Log.e(TAG, "No item [" + itemAttrib.id + "  -]" + e)
					}
				}
			}
		} catch (e: JSONException) {
			val fragmentError = newFragmentDialog(mFragment.activity!!.resources.getString(R.string.error_json_item),
					(mFragment.activity as BaseActivity?)!!, TypeError.ONE_OPTIONS)
			// DotaZoneBrain.mItems = new ArrayList<Item>();
			fragmentError.show(mFragment.activity!!.supportFragmentManager, "Item")
		}
	}

	fun createItem(result: String?) {
		try {
			val jsonResponse = JSONObject(result)
			val mainNode = jsonResponse["itemdata"] as JSONObject
			val keys: Iterator<*> = mainNode.keys()
			while (keys.hasNext()) {
				val key = keys.next() as String
				if (mainNode[key] is JSONObject) {
					val item = Item()
					val jsonChild = mainNode.getJSONObject(key)
					var componentsList: JSONArray? = null
					try {
						if (jsonChild.getString(ItemElementy.DNAME.toString().toLowerCase()) == "Dagon") {
							println()
						}
						item.id = jsonChild.getInt(ItemElementy.ID.toString().toLowerCase())
						item.description = jsonChild.getString(ItemElementy.DESC.toString().toLowerCase())
						item.mc = jsonChild.getString(ItemElementy.MC.toString().toLowerCase())
						item.isCreated = jsonChild.getBoolean(ItemElementy.CREATED.toString().toLowerCase())
						item.imageName = jsonChild.getString(ItemElementy.IMG.toString().toLowerCase())
						item.qual = jsonChild.getString(ItemElementy.QUAL.toString().toLowerCase())
						item.name = jsonChild.getString(ItemElementy.DNAME.toString().toLowerCase())
						item.lore = jsonChild.getString(ItemElementy.LORE.toString().toLowerCase())
						item.isCreated = jsonChild.getBoolean(ItemElementy.CREATED.toString().toLowerCase())
						item.cost = jsonChild.getInt(ItemElementy.COST.toString().toLowerCase())
						item.notes = jsonChild.getString(ItemElementy.NOTES.toString().toLowerCase())
						item.attribute = jsonChild.getString(ItemElementy.ATTRIB.toString().toLowerCase())
						item.cloudown = jsonChild.getString(ItemElementy.CD.toString().toLowerCase())
						try {
							componentsList = jsonChild.getJSONArray(ItemElementy.COMPONENTS.toString().toLowerCase())
						} catch (e: Exception) {

							// Trata os itens que não tem um componente.
							Log.i(TAG, "no improvement items[" + item.name + "  -]")
						}
						if (componentsList != null) {
							for (i in 0 until componentsList.length()) {
								item.components.add(componentsList.getString(i))
								Log.i(TAG, "Improvement items[" + item.name + "  -]")
							}
						}
						identifyItemCategoryAndAdd(item)
					} catch (e: JSONException) {
						Log.e(TAG, "No item [" + item.name + "  -]" + e)
					}
				}
			}
		} catch (e: JSONException) {
			val fragmentError = newFragmentDialog(mFragment.activity!!.resources.getString(R.string.error_json_item),
					(mFragment.activity as BaseActivity?)!!, TypeError.ONE_OPTIONS)
			// DotaZoneBrain.mItems = new ArrayList<Item>();
			fragmentError.show(mFragment.activity!!.supportFragmentManager, "Item")
		}
	}

	private fun identifyItemCategoryAndAdd(item: Item) {
		val pos = item.imageName.lastIndexOf('.')
		val id = mFragment.activity!!.resources
				.getIdentifier(item.imageName.substring(0, pos), "drawable", mFragment.activity!!.packageName)
		for (itemAttrib in mItemAttribs!!) {
			if (item.imageName == itemAttrib.id + ".png") {
				item.setmItemAttrib(itemAttrib)
				Log.i(TAG, "attribute item add-" + itemAttrib.id)
				break
			}
		}

		// Verifica se existem a imagem cadastrada no json.
		if (id > 0) {
			items.add(item)
			//mItems.add(item);
			Log.i(TAG, "adding Item [" + item.imageName + "]")
		}
	}

	override fun onPostExecute(result: String?) {
		super.onPostExecute(result)
		mProgressDialog.dismiss()
		mAdapterAction.initList(null)
	}

	override fun onPreExecute() {
		mProgressDialog = ProgressDialog(mFragment.activity)
		mProgressDialog.setMessage("Carregando...")
		mProgressDialog.show()
		mProgressDialog.setCancelable(false)
	}

	init {
		mRequest = RequestREST()
		mProgressDialog = ProgressDialog(mFragment.activity)
		mAdapterAction = mFragment as ItemFragment
	}
}

